relayer/src/kill_switch.rs

```diff
@@ -0,0 +1,176 @@
+    ) -> Self {
+        Self {
+            gear_api,
+            eth_api,
+            genesis_config,
+            proof_storage,
+            from_eth_block,
+            emergency_stop: false,
+        }
+    }
+
+    pub async fn run(&mut self) -> anyhow::Result<()> {
+        log::info!("Starting kill switch relayer");
+
+        loop {
+            let now = Instant::now();
+            let res = self.main_loop().await;
+
+            if let Err(err) = res {
+                log::error!("{}", err);
+            }
+
+            let main_loop_duration = now.elapsed();
+            if main_loop_duration < MIN_MAIN_LOOP_DURATION {
+                tokio::time::sleep(MIN_MAIN_LOOP_DURATION - main_loop_duration).await;
+            }
+        }
+    }
+
+    async fn main_loop(&mut self) -> anyhow::Result<()> {
+        if self.emergency_stop {
+            log::info!("Emergency stop triggered, skipping..");
+            return Ok(());
+        }
+
+        log::info!("Syncing authority set id");
+        loop {
+            let sync_steps = sync_authority_set_id(
+                &self.gear_api,
+                self.proof_storage.as_mut(),
+                self.genesis_config,
+                None,
+            )
+            .await?;
+            if sync_steps == 0 {
+                break;
+            } else {
+                log::info!("Synced {} authority set ids", sync_steps);
+            }
+        }
+
+        let last_finalized_block = self.eth_api.finalized_block_number().await?;
+
+        // Set the initial value for `from_eth_block` if it's not set yet.
+        if self.from_eth_block.is_none() {
+            self.from_eth_block = Some(last_finalized_block);
+        }
+
+        let from_eth_block = self.from_eth_block.expect("should be set above");
+        if last_finalized_block < from_eth_block {
+            log::info!(
+                "No new eth block, skipping.. last_processed_eth_block={}, last_finalized_block={}",
+                from_eth_block.saturating_sub(1),
+                last_finalized_block,
+            );
+            return Ok(());
+        }
+
+        let events = self
+            .eth_api
+            .fetch_merkle_roots_in_range(from_eth_block, last_finalized_block)
+            .await?;
+
+        for event in events {
+            if !self.compare_merkle_roots(&event).await? {
+                // Okay, we have a mismatch,
+                // that means for some reason the proof with incorrect merkle root was submitted to relayer contract.
+                // We need to generate the correct proof and submit it to the relayer contract.
+
+                log::info!("Proving merkle root presence");
+                let proof = self.generate_proof(event.block_number).await?;
+                log::info!("Proven merkle root presence");
+
+                log::info!("Submitting new proof to ethereum");
+                submit_proof_to_ethereum(&self.eth_api, proof).await?;
```

Comment:
Note that you'll also need to wait for tx to be finalized

Reply:
Fixed